### PR TITLE
Have z_shieldcoinbase use WalletTxBuilder

### PR DESCRIPTION
**Pull request by sellout**
NB: This does not yet support ZIP 317. That is in a separate PR.

Co-authored-by: Kris Nuttycombe <kris@nutty.land>

### CODE REVIEW

File: src/wallet/asyncrpcoperation_shieldcoinbase.h
Comment (nuttycom) [inline src/wallet/asyncrpcoperation_shieldcoinbase.h:43]: I don't understand why this is named `DUST_SIZE`. However, when making the subsequent ZIP 317 change, we should just use the ZIP 317 constant for transparent input size here.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (nuttycom) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp:178]: Thinking through this, nothing blocking; maybe improve the comment if making other changes:
* we have limited transparent UTXOs at this point, so we should be able to accurately compute the ZIP 317 fee here; I don't fully understand the "we don't know the amount at this point" comment.
* It would be possible to now make the recipient address of a `z_shieldcoinbase` transaction optional at the RPC level, using the change address for the account if not provided. However, this would slightly alter change pool determination, because we'd need to special-case the situation where *no* output pools are specified. For now it makes sense to me to just keep this as a payment. It's not critical here, but maybe we should add something to the wallet tx builder that allows the caller to force the change amount to be zero, ensuring that the inputs are fully consumed?
* I don't fully understand how the P2SH case factors in here (and elsewhere in ZIP 317 fee computation); for multisig P2SH the wallet would have to control all of the keys needed to make the signatures, for other P2SH scripts I guess it's possible that the wallet could store nonstandard spend scripts, and the low-level transaction builder can then spend those UTXOs correctly. My suspicion is that P2SH spends will usually underpay ZIP 317 fees for now, but that's okay because they'll still be close enough to be prioritized highly.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (str4d) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp:238]: Where were these locked? And if somewhere earlier inside `WalletTxBuilder`, why does this not also happen inside `WalletTxBuilder`?

File: qa/rpc-tests/remove_sprout_shielding.py
Comment (nuttycom) [inline qa/rpc-tests/remove_sprout_shielding.py:156]: Nonblocking: this change and some of those above are unnecessary here, it probably leaked in from my code, but it's fine.

File: qa/rpc-tests/wallet_shieldcoinbase.py
Comment (nuttycom) [inline qa/rpc-tests/wallet_shieldcoinbase.py:93]: This error message should probably make clear whether the units are ZAT or ZEC

**Review by str4d — CHANGES_REQUESTED** [review]
Removal of `COutput::Dummy` is blocking.

**Review by nuttycom — CHANGES_REQUESTED** [review]
Suggestion for change to the base transaction size to be 10000 bytes to accommodate Orchard; nothing else is blocking.

File: src/wallet/wallet.cpp
Comment (sellout) [reply src/wallet/wallet.cpp:7784]: Ah, good – I was just thinking that I wanted to make that change.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:89]: This change is about having shieldcoinbase only explicitly handle the behavior that is specific to it. The “can’t send to Sprout” policy is more general, and is handled in one place for all tx. This is only to check that shielding specifically doesn’t allow sending to transparent addresses. Although it could be clearer if it were encapsulated, like
```c++
bool RecipientIsTransparent(const PaymentAddress& address)
{
    return examine(address, …);
}

…

if (RecipientIsTransparent(destaddress.value())) {
    throw JSONRPCError(RPC_VERIFY_REJECTED, "Cannot shield coinbase output to a transparent address.");
}
```

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:60]: Yes – `AmountFromValue` which is used across the RPC to create `CAmount` from `UniValue` throws on `!MoneyRange` before we ever get here.

However, I would like to make the constructor for `CAmount` private and add a convenience constructor that returns optional or expected on `< -MAX_MONEY` or `> MAX_MONEY`, and then something like `class CMoney : CAmount` with a similar convenience constructor that tightens the check to `< 0`. Then we can know that values of these types are always in the correct range, avoiding asserts and checks elsewhere.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:238]: They are locked immediately after creating the `TransactionEffects`. I agree the asymmetry isn’t ideal, but we need to hold onto the lock until after we’ve sent the transaction, plus we need to handle the failure case where we never call `SendTransaction` (which this code is missing – fixing that!), so we can’t unlock in there (or at least not only in there, and so having all the unlocking in the same place is helpful).

In future, we’d like to hold the locks until after the expiry height, but this implementation matches the current semantics and avoids a bigger change for now.

File: src/wallet/rpcwallet.cpp
Comment (sellout) [reply src/wallet/rpcwallet.cpp:5234]: That’s a braino – it should have been `AllowFullyTransparent`, which is the only established meaning for LegacyCompat that makes sense here.

However, in the context of `z_shieldcoinbase`, `AllowLinkingAccountAddresses` might be the more sensible `LegacyCompat` meaning. I was thinking we could do a similar check as `z_sendmany` – if we’re shielding to a UA, `LegacyCompat` is `AllowRevealedSenders` (or `AllowFullyTransparent`, which has the same meaning here), and if there’s no UA, then `LegacyCompat` is `ALAA`.

But … it’s not quite the same, as (until #6467 is merged) `ALAA` only has meaning when you’re sending from a UA (or worse, `ANY_TADDR`), so conditionalizing on _any_ UA involved is pretty meaningless (we could just always default to `AllowRevealedSenders` and get the same semantics). Also, I’m assuming `LegacyCompat` for `z_sendmany` was added when UAs were added, so introducing it/privacyPolicies wasn’t a breaking change (it maintained the existing semantics for the tx that were possible before UAs), but here it would be a breaking change, as it changes the semantics for UA-involved tx that have already been established (in that previously using `ANY_TADDR` or a UA as the sender would imply `AllowLinkingAccountAddresses`).

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:178]: Ah, this is a good point.

My commit that’s blocked by this one introduces a `NetAmountPayment`, which generalizes `MergeToAddressRecipient` and does the fee & send_amount calculation in WalletTxBuilder for the remaining operations. But _maybe_ we can make a smaller change for now that calculates the fee for these outside WalletTxBuilder.

I think the thing we don’t know, when shielding to a UA, is which pool it’s going to. But … it can’t be transparent or Sprout, and both Sapling and Orchard pad to 2 outputs, so that’s enough info to calculate the fee. For mergetoaddress, it’s maybe a bit more complicated, as there’s also the chance that a UA recipient ends up going to a transparent receiver (e.g., merging Sprout to an Orchard+transparent UA), which would have a lower fee. But the existing changes I’m making for this already break `ResolvePayment` out of `ResolveInputsAndPayments`, so if I make that public, then mergetoaddress can also calculate its own fee correctly.

Anyway, I’ll have to consider it compared to my about to be pushed PR, so I think I’ll keep it as a separate PR, so the approach can be evaluated.

File: qa/rpc-tests/wallet_shieldcoinbase.py
Comment (sellout) [reply qa/rpc-tests/wallet_shieldcoinbase.py:93]: I added a `DisplayAmount` in a different branch … somewhere. Anyway, it’s basically `FormatAmount`, but prefixes with `CURRENCY_UNIT`. Later, I’d like it to also have thousands separators in it.

Since this is an existing error message, I’d like to do that in a separate PR, too, and do a `s/FormatAmount/DisplayAmount/g` more broadly.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:89]: Oh, sorry, I misread this comment. I thought this was talking about the change _in_ rpcwallet. I agree it’s better for it to live here instead of there.

File: src/wallet/rpcwallet.cpp
Comment (sellout) [reply src/wallet/rpcwallet.cpp:5234]: For now, I’ve made the simple change to use `AllowFullyTransparent`, but I‘d like some help figuring out the correct thing to do.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:60]: Oh, `CAmount` isn’t a class, of course. I was thinking of `CFeeRate`. So, nvm.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:238]: I could move `LockSpendable` out of `WalletTxBuilder` to here – I’d like to get them both _into_ WalletTxBuilder, but that isn’t feasible at the moment.

File: src/wallet/test/rpc_wallet_tests.cpp
Comment (sellout) [reply src/wallet/test/rpc_wallet_tests.cpp:1546]: Correct about the first one. No idea about the second. I restored it, using a ZTXOSelector that has nothing to select. It passes.

File: src/wallet/asyncrpcoperation_shieldcoinbase.h
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.h:43]: Renamed to `CTXIN_SPEND_P2PKH_SIZE`. Also found that it (and `CTXOUT_REGULAR_SIZE`) had duplicate definitions in rpcwallet.cpp … I’m guessing this PR intended to move them. So I removed those.

File: qa/rpc-tests/remove_sprout_shielding.py
Comment (sellout) [reply qa/rpc-tests/remove_sprout_shielding.py:156]: I’m inclined to leave them in … they improve the clarity & consistency, and some of them are entwined with actual changes.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:178]: The idea is that the `fee_` we've computed at this point won't have been able to take into account the actual size of the spend script that the low-level transaction builder ultimately uses to spend from the P2SH input. So we'll compute a fee that _may_ underestimate the ultimate transaction size, and consequently the ZIP 317 conventional fee for that tx.

File: src/wallet/rpcwallet.cpp
Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:5234]: In practice I don't think that there's much difference between `AllowFullyTransparent` and `AllowLinkingAccountAddresses` under `LegacyCompat` so this is fine for now.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:238]: Even moving them out, they’re in separate methods, because locking happens after preparation, and unlocking after approve/build/send. So, while it’s necessary to have the locking already, we want them to be in the right places so they just work when we switch to 2-stage tx.

Tangentially (and likely not for this PR) … z_shieldcoinbase already does preparation synchronously, since it returns some extra data that we get via preparation. Should _all_ of the async ops have sync preparation? Both for consistency and for 2-stage prep.

File: qa/rpc-tests/wallet_shieldcoinbase.py
Comment (nuttycom) [reply qa/rpc-tests/wallet_shieldcoinbase.py:93]: Created https://github.com/zcash/zcash/issues/6570 to track this.

**Review by nuttycom — DISMISSED** [review]
utACK 55ebf19c55d8291ec5d22d53c63232d8dc934317

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp]: I don't think that synchronous prep makes much difference here; the caller still has to check the async result for failures irrespective whether prep succeeded or failed.

Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp]: When you say “here”, what do you mean?
- in z_shieldcoinbase? In this case, we need to do it sync, just because z_shieldcoinbase return ancillary info that’s only available once we prepare.
- in general, before 2-stage? I agree it doesn’t make a lot of difference, which makes me inclined to do sync prep (after this release) simply because with the consistency, we can maybe move more logic to `wallet_tx_builder` `async…_common`, etc. rather than duplicating between calls; and also because that’s roughly how 2-stage tx will work (right?)

**Review by str4d — APPROVED** [review]
utACK 55ebf19c55d8291ec5d22d53c63232d8dc934317

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:187]: Ah, building `CTxIn`/`Out` is pretty trivial. So, WalletTxBuilder currently does that just to calculate the fee from the serialization and throw it away. So, I think we’d get that right no matter where we do it (as long as we have something like UTXOs and not just a count).

**Review by nuttycom — DISMISSED** [review]
re-utACK 1ba0843845bd9edfb50309f9f1e80cda0d7e9217

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (daira) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:60]: +1 on adding wrappers for signed and unsigned amounts, even if they have to be separate from CAmount so that we can migrate gradually. We'll need to do that anyway for ZSAs.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (daira) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp:71]: Non-blocking: reminder that the convention is to only indent a match 4 spaces, not 8.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (daira) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp:143]:
```suggestion
    size_t estimatedTxSize = MIN_TX_COST; // per ZIP 401 (https://zips.z.cash/zip-0401#specification)
```
(defined in `mempool_limit.h`). But ZIP 401 doesn't specify that we use this for tx construction, so it's a bit confusing to be referencing that.

If I understand correctly, this is an initial estimate that is going to be refined in subsequent iterations, so it doesn't really matter much what we start with? In which case using MIN_TX_COST is fine, but please say something like that in the comment.

Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:143]: It doesn’t really get “refined” so much as increased as new inputs are added. So we’re starting with the base being the most expensive recipient, then bump the size as we add inputs.

I’ll fix the comment (and use the constant).

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (sellout) [reply src/wallet/asyncrpcoperation_shieldcoinbase.cpp:71]: Ah, this one was a copy/paste mistake from rpcwallet, where it was nested another level.

**Review by nuttycom — APPROVED** [review]
re-utACK https://github.com/zcash/zcash/commit/f4933793a07ac6e0f946cceb81e100704f78a2fa

**Review by nuttycom — APPROVED** [review]
re-re-utACK https://github.com/zcash/zcash/commit/1ba0843845bd9edfb50309f9f1e80cda0d7e9217

File: src/wallet/asyncrpcoperation_shieldcoinbase.h
Comment (daira) [inline src/wallet/asyncrpcoperation_shieldcoinbase.h:28]: Why 9165? That's the size of a fully shielded Orchard transaction with 2 actions. The marginal size per receiver is much less.

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp
Comment (daira) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp:30]:
```suggestion
#include "mempool_limit.h"
#include "miner.h"
```

**Pull request merged**
